src/caffe/layers/detectnet_transform_layer.cpp

```diff
@@ -0,0 +1,929 @@
+        Size(images.width(), images.height())
+    ));
+  }
+
+  return result;
+}
+
+
+template<typename Dtype>
+vector<vector<typename DetectNetTransformationLayer<Dtype>::BboxLabel> >
+DetectNetTransformationLayer<Dtype>::blobToLabels(
+    const Blob<Dtype>& labels
+) const {
+  vector<vector<BboxLabel > > result; result.reserve(labels.num());
+
+  for (size_t iLabel = 0; iLabel != labels.num(); ++iLabel) {
+    const Dtype* source = &labels.cpu_data()[
+        labels.offset(iLabel, 0, 0, 0)
+    ];
+
+    size_t numOfBbox = static_cast<size_t>(source[0]);
+    size_t bboxLen = static_cast<size_t>(source[1]);
+    CHECK_EQ(bboxLen, sizeof(BboxLabel) / sizeof(Dtype));
+    CHECK_LE(numOfBbox, labels.height());
+    // exclude header
+    source += bboxLen;
+    // convert label into typed struct:
+    result.push_back(vector<BboxLabel>(
+        reinterpret_cast<const BboxLabel*>(source),
+        reinterpret_cast<const BboxLabel*>(source + bboxLen * numOfBbox)
+    ));
+  }
+
+  return result;
+}
+
+
+template<typename Dtype>
+struct toDtype : public std::unary_function<float, Dtype> {
+  Dtype operator() (const Vec<Dtype, 1>& value) { return value(0); }
+};
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::matToBlob(
+    const Mat3v& source,
+    Dtype* destination
+) const {
+  std::vector<Mat1v> channels;
+  split(source, channels);
+
+  size_t offset = 0;
+  for (size_t iChannel = 0; iChannel != channels.size(); ++iChannel) {
+    const Mat1v& channel = channels[iChannel];
+
+    std::transform(
+        channel.begin(),
+        channel.end(),
+        &destination[offset],
+        toDtype<Dtype>()
+    );
+
+    offset += channel.total();
+  }
+}
+
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::matsToBlob(
+    const vector<Mat3v>& _source,
+    Blob<Dtype>& _dest
+) const {
+  for (size_t iImage = 0; iImage != _source.size(); ++iImage) {
+    Dtype* destination = &_dest.mutable_cpu_data()[
+        _dest.offset(iImage, 0, 0, 0)
+    ];
+    const Mat3v& source = _source[iImage];
+    matToBlob(source, destination);
+  }
+}
+
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::pixelMeanSubtraction(
+    Mat3v& source
+) const {
+  source -= data_mean_;
+}
+
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::channelMeanSubtraction(
+    Mat3v& source
+) const {
+  vector<Mat1f> channels;
+  split(source, channels);
+  for(size_t iChannel = 0; iChannel != channels.size(); ++iChannel) {
+    channels[iChannel] -= mean_values_[iChannel];
+  }
+  merge(channels, source);
+}
+
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::meanSubtract(Mat3v& source) const {
+  if (t_param_.has_mean_file()) {
+    pixelMeanSubtraction(source);
+  } else if (t_param_.mean_value_size() != 0) {
+    channelMeanSubtraction(source);
+  }
+}
+
+
+template<typename Dtype>
+Dtype DetectNetTransformationLayer<Dtype>::randDouble() {
+  CHECK(rng_);
```

Comment:
clarification: I don't think any code path where this test could be useful.
If the raw memory allocation failed or if constructing the `shared_ptr` failed, an exception should have been thrown already.